packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -608,6 +609,45 @@ class NavigatorObserver {
   void didStopUserGesture() { }
 }
 
+/// An inherited widget to host a hero controller.
+///
+/// This class should not be used directly. The [MaterialApp] and [CupertinoApp]
+/// use this class to host the [HeroController], and they should be the only
+/// exception to use this class. If you want to subscribe your own
+/// [HeroController], use the [Navigator.observers] instead.
+///
+/// The hosted hero controller will be picked up by the navigator in the
+/// [child] subtree. Once a navigator picks up this controller, the navigator
+/// will bar any navigator below its subtree from receiving this controller.
+///
+/// See also:
+///
+///  * [Navigator.observers], which is the standard way of providing a
+///    [HeroController].
+class HeroControllerScope extends InheritedWidget {
+  /// Creates a widget to host the input [controller].
+  const HeroControllerScope({
```

Comment:
would be clearer to have two constructors, one that requires that controller be non-null, and one that doesn't have a controller argument, with a name like `HeroControllerScope.none` or something.

Reply:
This is a temporary workaround for router api, eventually we want to build a more flexible system. I think we should not further improve this api?